### PR TITLE
Exhaustiveness: Improve complexity on some wide matches

### LINKED ISSUE #118437: exhaustiveness checking hangs forever when compiling pulsar-rs
```bash
git clone https://github.com/streamnative/pulsar-rs
cd pulsar-rs
cargo +nightly check
```

Then it hangs forever

bisect leads to https://github.com/rust-lang/rust/commit/ee80c8d0a8bc63b69f68216c5d37f9ab837eedd0
cc @Nadrieril @cjgillot 

searched nightlies: from nightly-2023-11-26 to nightly-2023-11-29
regressed nightly: nightly-2023-11-27
searched commit range: https://github.com/rust-lang/rust/compare/f5dc2653fdd8b5d177b2ccbd84057954340a89fc...6cf088810f66fff15d05bf7135c5f5888b7c93b4
regressed commit: https://github.com/rust-lang/rust/commit/ee80c8d0a8bc63b69f68216c5d37f9ab837eedd0

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.7</summary>

Host triple: aarch64-apple-darwin
Reproduce with:
```bash
cargo bisect-rustc --start 2023-11-26 --script ./test.sh 
```
</details>

bisect script:
```
#!/bin/sh

set -ex

timeout 40s cargo check
```

### PR DESCRIPTION
**Pull request by Nadrieril**
https://github.com/rust-lang/rust/issues/118437 revealed an exponential case in exhaustiveness checking. While [exponential cases are unavoidable](https://compilercrim.es/rust-np/), this one only showed up after my https://github.com/rust-lang/rust/pull/117611 rewrite of the algorithm. I remember anticipating a case like this and dismissing it as unrealistic, but here we are :').

The tricky match is as follows:
```rust
match command {
    BaseCommand { field01: true, .. } => {}
    BaseCommand { field02: true, .. } => {}
    BaseCommand { field03: true, .. } => {}
    BaseCommand { field04: true, .. } => {}
    BaseCommand { field05: true, .. } => {}
    BaseCommand { field06: true, .. } => {}
    BaseCommand { field07: true, .. } => {}
    BaseCommand { field08: true, .. } => {}
    BaseCommand { field09: true, .. } => {}
    BaseCommand { field10: true, .. } => {}
    // ...20 more of the same

    _ => {}
}
```

To fix this, this PR formalizes a concept of "relevancy" (naming is hard) that was already used to decide what patterns to report. Now we track it for every row, which in wide matches like the above can drastically cut on the number of cases we explore. After this fix, the above match is checked with linear-many cases instead of exponentially-many.

Fixes https://github.com/rust-lang/rust/issues/118437

r? @cjgillot

### CODE REVIEW

**Comment by lqd** [comment]
Out of curiosity, how much does this PR improve things on #118437, and how does that compare to the situation prior to #117611?

**Comment by Nadrieril** [comment]
On #118437, this goes from "takes so long that the playground timeouts" to "compiles instantly".

Compared to before the one-pass rewrite, my guess is that it's comparable in performance. Our benchmarks don't have a lot of variety on matches so it's hard to measure. The rewrite was more for code health than for performance honestly, the difference would only be noticeable for somewhat complex matches. Our benchmarks barely noticed.

**Comment by Nadrieril** [comment]
Let's see how this fares:

@bors try @rust-timer queue

**Comment by lqd** [comment]
> On #118437, this goes from "takes so long that the playground timeouts" to "compiles instantly".

So if I understand correctly, #118437 also compiled instantly prior to #117611.

**Comment by Nadrieril** [comment]
> So if I understand correctly, #118437 also compiled instantly prior to #117611.

Indeed! You can try it here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1f9de10106c7eec6d18773c9804e614f . On stable is the pre-#117611 behavior; on nightly is the current exponential behavior.

Between pre-#117611 and post-this-PR, the tradeoff is roughly: pre-#117611 can check some cases several times, which post-this-PR never does; conversely, there are cases that post-this-PR considers distinct but pre-#117611 does in one go.

**Comment by apiraino** [comment]
I'm going to pre-emptively nominated this for beta backport in case the review and merge make it to **Dec, 21st**. 

Ref. [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/release.20schedule.20change.20for.201.2E75) where T-infra announces that the beta cut will happen a bit earlier in this release cycle

@rustbot label +beta-nominated

File: compiler/rustc_pattern_analysis/src/usefulness.rs [hunk 1]
```diff
@@ -300,6 +300,74 @@
...
+//! When we then specialize with `North`, we have two arms left:
+//!
+//! ```ignore(partial code)
+//!     (_) => {} // arm 1
+//!     (50..) => {} // arm 2
+//! ```
+//!
+//! Because `Missing` only matches wildcard rows, specializing with `Missing` is guaranteed to
+//! result in a subset of the rows obtained from specializing with anything else. This means that
+//! any row with a wildcard found useful when specializing with anything else would also be found
```
Comment (cjgillot) [inline compiler/rustc_pattern_analysis/src/usefulness.rs:335]:
```suggestion
//! any row with a wildcard found useful when specializing with any non-`Missing` constructor would also be found
```
?

File: compiler/rustc_pattern_analysis/src/usefulness.rs [hunk 2]
```diff
@@ -300,6 +300,74 @@
...
+//! enum Direction { North, South, East, West }
+//! # let wind = (Direction::North, 0u8);
+//! match wind {
+//!     (Direction::North, _) => {} // arm 1
+//!     (_, 50..) => {} // arm 2
+//! }
+//! ```
+//!
+//! Remember that we represent the "everything else" cases with [`Constructor::Missing`]. When we
+//! specialize with `Missing` in the first column, we have one arm left:
```
Comment (cjgillot) [inline compiler/rustc_pattern_analysis/src/usefulness.rs:317]: This is the first time in this comment that you mention `Missing`. This makes the "remember" verb a bit odd.

File: compiler/rustc_pattern_analysis/src/usefulness.rs [hunk 3]
```diff
@@ -300,6 +300,74 @@
...
+//!
+//! Because `Missing` only matches wildcard rows, specializing with `Missing` is guaranteed to
+//! result in a subset of the rows obtained from specializing with anything else. This means that
+//! any row with a wildcard found useful when specializing with anything else would also be found
+//! useful in the `Missing` case. In our example, after specializing with `North` here we will not
+//! gain new information regarding the usefulness of arm 2 or of the fake wildcard row used for
+//! exhaustiveness. This allows us to skip cases.
+//!
+//! When specializing, if there is a `Missing` case we call the other constructors "irrelevant".
+//! When there is no `Missing` case there are no irrelevant constructors.
```
Comment (cjgillot) [inline compiler/rustc_pattern_analysis/src/usefulness.rs:341]: So in the example, the constructor `North` is irrelevant?

File: compiler/rustc_pattern_analysis/src/usefulness.rs [hunk 4]
```diff
@@ -300,6 +300,74 @@
...
+//!     (_, true, _) => {}
+//! }
+//! ```
+//!
+//! In this example we will skip the `(true, true, _)` case entirely. Thus `(true, true, false)`
+//! will not be reported as missing. In fact we go further than this: we deliberately do not report
+//! any cases that are irrelevant for the fake wildcard row. For example, in `match ... { (true,
+//! true) => {} }` we will not report `(true, false)` as missing. This was a deliberate choice made
+//! early in the development of rust; it so happens that it is beneficial for performance reasons
+//! too.
```
Comment (cjgillot) [inline compiler/rustc_pattern_analysis/src/usefulness.rs:367]: I really have a hard time understanding the logic here. Could you give an example of the runtime for the first example, like you did in other paragraphs?

Comment (Nadrieril) [reply compiler/rustc_pattern_analysis/src/usefulness.rs:367]: You're right, I've got to spell this out in more detail

File: compiler/rustc_pattern_analysis/src/usefulness.rs (hunk 3 above)
Comment (Nadrieril) [reply compiler/rustc_pattern_analysis/src/usefulness.rs]: That was poorly phrased, relevancy really depends on the rows. I reworked the explanation entirely

**Comment by Nadrieril** [comment]
I reworked the explanation entirely, I hope the new one answers your questions

**Comment by cjgillot** [comment]
Thanks!
@bors r+

**Comment by apiraino** [comment]
@Nadrieril thank you :slightly_smiling_face:  I see the PR has been merged with some minuses in the perf. report. Do these results need a comment? Will they be triaged during the usual weekly perf. triage?

**Comment by Nadrieril** [comment]
Perf feels ok to me; it's a stress-test and I'm not surprised it was affected by this change

**Comment by apiraino** [comment]
Awesome, then I'll mark this as perf. triaged (I _think_ this is the correct process, in case it's not please anyone feel free to jump in and correct me :slightly_smiling_face: )

@rustbot label -perf-regression +perf-regression-triaged

**Comment by Kobzol** [comment]
@apiraino I think that normally we just apply the `perf-regression-triaged` label, without removing the `perf-regression` label (check the example command in the [perf comment](https://github.com/rust-lang/rust/pull/118796#issuecomment-1868573445)).

**Comment by apiraino** [comment]
Beta backport approved as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-12-28/near/410306503)

@rustbot label +beta-accepted

**Comment by Nadrieril** [comment]
I seem to understand that someone will swoop in and backport this along with a bunch of other PRs at some point? Ping me if this doesn't backport cleanly.

**Pull request merged**